consensus/src/pending_order_votes.rs

```diff
@@ -0,0 +1,237 @@
+    VoteAdded(u128),
+    /// This block has just been certified after adding the vote.
+    NewLedgerInfoWithSignatures(LedgerInfoWithSignatures),
+    /// There might be some issues adding a vote
+    ErrorAddingVote(VerifyError),
+    /// Error happens when aggregating signature
+    ErrorAggregatingSignature(VerifyError),
+    /// The author of the order vote is unknown
+    UnknownAuthor(Author),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum OrderVoteStatus {
+    EnoughVotes(LedgerInfoWithSignatures),
+    NotEnoughVotes(LedgerInfoWithPartialSignatures),
+}
+
+/// A PendingVotes structure keep track of order votes for the last few rounds
+pub struct PendingOrderVotes {
+    /// Maps LedgerInfo digest to associated signatures (contained in a partial LedgerInfoWithSignatures).
+    /// Order vote status stores caches the information on whether the votes are enough to form a QC.
+    li_digest_to_votes: HashMap<HashValue /* LedgerInfo digest */, OrderVoteStatus>,
+}
+
+impl PendingOrderVotes {
+    /// Creates an empty PendingOrderVotes structure
+    pub fn new() -> Self {
+        Self {
+            li_digest_to_votes: HashMap::new(),
+        }
+    }
+
+    /// Add a vote to the pending votes
+    // TODO: Should we add any counters here?
+    pub fn insert_order_vote(
+        &mut self,
+        order_vote: &OrderVote,
+        validator_verifier: &ValidatorVerifier,
+    ) -> OrderVoteReceptionResult {
+        // derive data from order vote
+        let li_digest = order_vote.ledger_info().hash();
+
+        // obtain the ledger info with signatures associated to the order vote's ledger info
+        let status = self.li_digest_to_votes.entry(li_digest).or_insert_with(|| {
+            // if the ledger info with signatures doesn't exist yet, create it
+            OrderVoteStatus::NotEnoughVotes(LedgerInfoWithPartialSignatures::new(
+                order_vote.ledger_info().clone(),
+                PartialSignatures::empty(),
+            ))
+        });
+
+        match status {
+            OrderVoteStatus::EnoughVotes(li_with_sig) => {
+                // we already have enough votes for this ledger info
+                OrderVoteReceptionResult::NewLedgerInfoWithSignatures(li_with_sig.clone())
+            },
+            OrderVoteStatus::NotEnoughVotes(li_with_sig) => {
+                // we don't have enough votes for this ledger info yet
+                let validator_voting_power =
+                    validator_verifier.get_voting_power(&order_vote.author());
+                if validator_voting_power.is_none() {
+                    warn!(
+                        "Received order vote from an unknown author: {}",
+                        order_vote.author()
+                    );
+                    return OrderVoteReceptionResult::UnknownAuthor(order_vote.author());
+                }
+                let validator_voting_power = validator_voting_power.unwrap();
+
+                if validator_voting_power == 0 {
```

Comment:
Should we just return error here? Does our signature support 0 voting power?

Reply:
Actually, we have the same if statement for pending votes as well. We don't return an error there if validator_voting_power is 0. The validator_verifier code verifies the aggregate signature, and checks if the sum of voting power of all the signed validators is greater than the threshold. It seems to be okay even if a validator has 0 voting power.